libavformat/rtcenc.c

```diff
@@ -0,0 +1,2226 @@
+            if (ret > 0 && (ret < 13 || buf[0] <= 19 || buf[0] >= 64))
+                continue;
+
+            /* Got DTLS response successfully. */
+            if (ret > 0) {
+                res_size = ret;
+                ctx->dtls_should_reset_timer = 1;
+                break;
+            }
+
+            /* Fatal error or timeout. */
+            if (ret != AVERROR(EAGAIN)) {
+                av_log(s1, AV_LOG_ERROR, "DTLS: Read response failed, loop=%d, content=%d, handshake=%d\n",
+                    loop, req_ct, req_ht);
+                return ret;
+            }
+
+            /* DTLSv1_handle_timeout is called when a DTLS handshake timeout expires. If no timeout
+             * had expired, it returns 0. Otherwise, it retransmits the previous flight of handshake
+             * messages and returns 1. If too many timeouts had expired without progress or an error
+             * occurs, it returns -1. */
+            r0 = DTLSv1_handle_timeout(dtls);
+            if (!r0) {
+                av_usleep(DTLS_SSL_TIMER_BASE + ctx->dtls_arq_timeout * 1000);
+                continue; /* no timeout had expired. */
+            }
+            if (r0 != 1) {
+                r1 = SSL_get_error(dtls, r0);
+                av_log(s1, AV_LOG_ERROR, "DTLS: Handle timeout, loop=%d, content=%d, handshake=%d, r0=%d, r1=%d\n",
+                    loop, req_ct, req_ht, r0, r1);
+                return AVERROR(EIO);
+            }
+
+            ctx->dtls_arq_packets++;
+            break;
+        }
+    }
+
+    /* Trace the response packet, feed to SSL. */
+    BIO_reset(bio_in);
+    openssl_state_trace(ctx, buf, res_size, 1, r0, SSL_ERROR_NONE);
+    res_ct = res_size > 0 ? buf[0]: 0;
+    res_ht = res_size > 13 ? buf[13] : 0;
+    if ((r0 = BIO_write(bio_in, buf, res_size)) <= 0) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: Feed response failed, loop=%d, content=%d, handshake=%d, size=%d, r0=%d\n",
+            loop, res_ct, res_ht, res_size, r0);
+        return AVERROR(EIO);
+    }
+
+    return ret;
+}
+
+/**
+ * DTLS handshake with server, as a client in active mode, using openssl.
+ *
+ * This function initializes the SSL context as the client role using OpenSSL and
+ * then performs the DTLS handshake until success. Upon successful completion, it
+ * exports the SRTP material key.
+ *
+ * @return 0 if OK, AVERROR_xxx on error
+ */
+static int dtls_context_handshake(DTLSContext *ctx)
+{
+    int ret, loop;
+    SSL_CTX *dtls_ctx = NULL;
+    SSL *dtls = NULL;
+    const char* dst = "EXTRACTOR-dtls_srtp";
+    BIO *bio_in = NULL, *bio_out = NULL;
+    int64_t starttime = av_gettime();
+    void *s1 = ctx->log_avcl;
+
+#if OPENSSL_VERSION_NUMBER < 0x10002000L /* OpenSSL v1.0.2 */
+    dtls_ctx = SSL_CTX_new(DTLSv1_method());
+#else
+    dtls_ctx = SSL_CTX_new(DTLS_client_method());
+#endif
+    if (!dtls_ctx) {
+        ret = AVERROR(ENOMEM);
+        goto end;
+    }
+
+    if (!ctx->udp_uc) {
```

Comment:
Derek Buitenhuis:

```text
> +    if (!ctx->udp_uc) {
> +        av_log(s1, AV_LOG_ERROR, "DTLS: No UDP context\n");
```

Kind of odd this would be possible?